### PR TITLE
Add Scrypt PKDF

### PR DESCRIPTION
**Pull request by shamatar**
- Add implementations of Scrypt PKDF
- Add test for Scrypt from RFC 

Fixes #

Checklist:
- [x] Pull request against a `develop` branch.
- [x] Correct file headers (see CONTRIBUTING.md).
- [x] Formatted with [SwiftFormat](https://github.com/nicklockwood/SwiftFormat).
- [x] Tests added.

Changes proposed in this pull request:
- 
- 
-

### CODE REVIEW

**Comment by krzyzanowskim** [comment]
it needs more work. looks like major cleanup is needed, starting ith typealias, that's unnecessary

**Comment by shamatar** [comment]
Clarification needed for "cleanups". Scrypt is included under the same namespace as PKDF2, Salsa is separate with merely two public functions.

**Comment by krzyzanowskim** [comment]
CI builds fails, lets start with that.
Separate Salsa and Scrypt in two PRs please. I'll review it separately.

**Comment by shamatar** [comment]
Hello @krzyzanowskim 

Yes, I should update the pull request all together, the naive realization using arrays and slices is very slow in performance, I've made another one using manual memory management, will post it next week.

Sincerely, Alex

**Comment by krzyzanowskim** [comment]
ping

**Comment by krzyzanowskim** [comment]
pong

File: Sources/CryptoSwift/PKCS/Scrypt.swift [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    //    DesiredKeyLen:
+    
+    public struct Scrypt {
+        public enum Error: Swift.Error {
+            case invalidInput
+            case derivedKeyTooLong
+        }
+        
+        private let salt: Array<UInt8> // S
+        private let password: Array<UInt8>
```
[the remark this answers is not in the document]
Comment (shamatar) [reply Sources/CryptoSwift/PKCS/Scrypt.swift]: Was following an example of other KDFs, where it's not a SecureBytes. Also would have to copy anyway for initial and final runs of PKDF2

File: Sources/CryptoSwift/PKCS/Scrypt.swift [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            }
+            kdf = try CryptoSwift.PKCS5.PBKDF2(password: self.password, salt: B, iterations: 1, keyLength: dkLen, variant: .sha256)
+            let ret = try kdf.calculate()
+            return Array(ret)
+        }
+    }
+}
+
+extension PKCS5.Scrypt {
+    static func blockCopy(destination: inout Array<UInt32>, source: Array<UInt32>, n: Int) {
```
[the remark this answers is not in the document]
Comment (shamatar) [reply Sources/CryptoSwift/PKCS/Scrypt.swift]: resolved in update

File: Sources/CryptoSwift/PKCS/Scrypt.swift [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+//  https://www.ietf.org/rfc/rfc7914.txt
+//
+
+#if canImport(Darwin)
+import Darwin
+#else
+import Glibc
+#endif
+
+public extension PKCS5 {
```
[the remark this answers is not in the document]
Comment (shamatar) [reply Sources/CryptoSwift/PKCS/Scrypt.swift]: Moved to Scrypt class, now not PKCS5

**Comment by shamatar** [comment]
Hello @krzyzanowskim 

I've updated an implementation that now gives a speed on par with `libsodium` when build for release. Manual inlining specifiers are helpful. Also, there is no more Salsa exposed as I was implementing it only for Scrypt purposes and not for steam cyphers. Implementing Salsa is most likely unnecessary (ChaCha20 is already there) and would require to change a codestyle to be similar to a ChaCha20 (different types, no raw memory play)

Sincerely, Alex

**Comment by krzyzanowskim** [comment]
Diff doesn't look right. There are unnecessary changes in the diff. What happened?

**Comment by shamatar** [comment]
Looks like file permissions got changed. I’ll try to rebase again

**Pull request closed without merging**